File: src-tauri/src/commands/explorer.rs
use std::path::PathBuf;

use crate::models::explorer::{ExplorerNode, FilePreview};
use crate::services::explorer;

#[tauri::command]
pub async fn pick_workspace_folder(show_hidden_files: Option<bool>) -> Result<Option<ExplorerNode>, String> {
    let show_hidden_files = show_hidden_files.unwrap_or(false);
    let selected_directory =
        tauri::async_runtime::spawn_blocking(|| rfd::FileDialog::new().pick_folder())
            .await
            .map_err(|error| error.to_string())?;

    let Some(path) = selected_directory else {
        return Ok(None);
    };

    let root = tauri::async_runtime::spawn_blocking(move || {
        explorer::build_workspace_root(&path, show_hidden_files)
    })
    .await
    .map_err(|error| error.to_string())??;

    Ok(Some(root))
}

#[tauri::command]
pub async fn scan_workspace_folder(
    root_path: String,
    show_hidden_files: Option<bool>,
) -> Result<ExplorerNode, String> {
    let root_path = PathBuf::from(root_path);
    let show_hidden_files = show_hidden_files.unwrap_or(false);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::build_workspace_root(&root_path, show_hidden_files)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn read_workspace_directory(
    root_path: String,
    directory_path: String,
    show_hidden_files: Option<bool>,
) -> Result<Vec<ExplorerNode>, String> {
    let root_path = PathBuf::from(root_path);
    let directory_path = PathBuf::from(directory_path);
    let show_hidden_files = show_hidden_files.unwrap_or(false);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::read_directory_children(&root_path, &directory_path, show_hidden_files)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn read_workspace_file(path: String) -> Result<FilePreview, String> {
    let file_path = PathBuf::from(path);

    tauri::async_runtime::spawn_blocking(move || explorer::read_workspace_file(&file_path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn create_markdown_file(
    root_path: String,
    selected_path: Option<String>,
    file_name: String,
) -> Result<ExplorerNode, String> {
    let root_path = PathBuf::from(root_path);
    let selected_path = selected_path.map(PathBuf::from);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::create_markdown_file(&root_path, selected_path.as_deref(), &file_name)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn create_workspace_directory(
    root_path: String,
    selected_path: Option<String>,
    directory_name: String,
) -> Result<ExplorerNode, String> {
    let root_path = PathBuf::from(root_path);
    let selected_path = selected_path.map(PathBuf::from);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::create_workspace_directory(&root_path, selected_path.as_deref(), &directory_name)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn write_workspace_file(path: String, content: String) -> Result<(), String> {
    let file_path = PathBuf::from(path);

    tauri::async_runtime::spawn_blocking(move || explorer::write_workspace_file(&file_path, &content))
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn rename_workspace_node(
    root_path: String,
    target_path: String,
    new_name: String,
) -> Result<(), String> {
    let root_path = PathBuf::from(root_path);
    let target_path = PathBuf::from(target_path);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::rename_workspace_node(&root_path, &target_path, &new_name)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn delete_workspace_node(root_path: String, target_path: String) -> Result<(), String> {
    let root_path = PathBuf::from(root_path);
    let target_path = PathBuf::from(target_path);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::delete_workspace_node(&root_path, &target_path)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn move_workspace_node(
    root_path: String,
    source_path: String,
    destination_directory: String,
) -> Result<(), String> {
    let root_path = PathBuf::from(root_path);
    let source_path = PathBuf::from(source_path);
    let destination_directory = PathBuf::from(destination_directory);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::move_workspace_node(&root_path, &source_path, &destination_directory)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn copy_workspace_node(
    root_path: String,
    source_path: String,
    destination_directory: String,
) -> Result<(), String> {
    let root_path = PathBuf::from(root_path);
    let source_path = PathBuf::from(source_path);
    let destination_directory = PathBuf::from(destination_directory);

    tauri::async_runtime::spawn_blocking(move || {
        explorer::copy_workspace_node(&root_path, &source_path, &destination_directory)
    })
    .await
    .map_err(|error| error.to_string())?
}

File: src/invoke/explorer.ts
import { invoke } from '@tauri-apps/api/core';
import type { ExplorerNode, FilePreview } from '@/components/explorer/types';

/** Opens a native folder picker and scans the selected workspace. */
export async function pickWorkspaceFolder(opts: {
	showHiddenFiles: boolean;
	sort?: boolean;
}): Promise<ExplorerNode | null> {
	return invoke<ExplorerNode | null>('pick_workspace_folder', {
		showHiddenFiles: opts.showHiddenFiles,
		sort: opts.sort ?? true,
	});
}

/** Scans a workspace folder into the explorer tree. */
export async function scanWorkspaceFolder(opts: {
	rootPath: string;
	showHiddenFiles?: boolean;
	sort?: boolean;
}): Promise<ExplorerNode> {
	return invoke<ExplorerNode>('scan_workspace_folder', {
		rootPath: opts.rootPath,
		showHiddenFiles: opts.showHiddenFiles ?? false,
		sort: opts.sort ?? true,
	});
}

/** Reads the children of a directory node. */
export async function readWorkspaceDirectory(opts: {
	directoryPath: string;
	rootPath: string;
	showHiddenFiles: boolean;
	sort?: boolean;
}): Promise<ExplorerNode[]> {
	return invoke<ExplorerNode[]>('read_workspace_directory', {
		directoryPath: opts.directoryPath,
		rootPath: opts.rootPath,
		showHiddenFiles: opts.showHiddenFiles,
		sort: opts.sort ?? true,
	});
}

/** Reads a workspace file with encoding detection. */
export async function readWorkspaceFile(opts: {
	path: string;
}): Promise<FilePreview> {
	return invoke<FilePreview>('read_workspace_file', {
		path: opts.path,
	});
}

/** Creates a new markdown file in the workspace. */
export async function createMarkdownFile(opts: {
	fileName: string;
	rootPath: string;
	selectedPath: string | null;
}): Promise<ExplorerNode> {
	return invoke<ExplorerNode>('create_markdown_file', {
		fileName: opts.fileName,
		rootPath: opts.rootPath,
		selectedPath: opts.selectedPath,
	});
}

/** Creates a new directory in the workspace. */
export async function createWorkspaceDirectory(opts: {
	directoryName: string;
	rootPath: string;
	selectedPath: string | null;
}): Promise<ExplorerNode> {
	return invoke<ExplorerNode>('create_workspace_directory', {
		directoryName: opts.directoryName,
		rootPath: opts.rootPath,
		selectedPath: opts.selectedPath,
	});
}

/** Writes content to a workspace file. */
export async function writeWorkspaceFile(opts: {
	content: string;
	path: string;
}): Promise<void> {
	return invoke('write_workspace_file', {
		content: opts.content,
		path: opts.path,
	});
}

/** Renames a file or directory in the workspace. */
export async function renameWorkspaceNode(opts: {
	newName: string;
	rootPath: string;
	targetPath: string;
}): Promise<void> {
	return invoke('rename_workspace_node', {
		newName: opts.newName,
		rootPath: opts.rootPath,
		targetPath: opts.targetPath,
	});
}

/** Deletes a file or directory from the workspace. */
export async function deleteWorkspaceNode(opts: {
	rootPath: string;
	targetPath: string;
}): Promise<void> {
	return invoke('delete_workspace_node', {
		rootPath: opts.rootPath,
		targetPath: opts.targetPath,
	});
}

/** Copies a file or directory within the workspace. */
export async function copyWorkspaceNode(opts: {
	destinationDirectory: string;
	rootPath: string;
	sourcePath: string;
}): Promise<void> {
	return invoke('copy_workspace_node', {
		destinationDirectory: opts.destinationDirectory,
		rootPath: opts.rootPath,
		sourcePath: opts.sourcePath,
	});
}

/** Moves a file or directory within the workspace. */
export async function moveWorkspaceNode(opts: {
	destinationDirectory: string;
	rootPath: string;
	sourcePath: string;
}): Promise<void> {
	return invoke('move_workspace_node', {
		destinationDirectory: opts.destinationDirectory,
		rootPath: opts.rootPath,
		sourcePath: opts.sourcePath,
	});
}
